File: frontend/src/services/study-planner.service.ts
/**
 * Study Planner service — CRUD operations for study task management.
 *
 * Communicates with the backend StudyPlannerController.
 *
 * @see backend/src/main/java/com/devlaunch/controller/StudyPlannerController.java
 * @author DevLaunch
 */

import apiClient from '../api/client';
import { STUDY_PLANNERS } from '../api/endpoints';
import type {
  StudyPlannerResponse,
  CreateStudyPlannerRequest,
  UpdateStudyPlannerRequest,
} from '../types/study-planner';

export const studyPlannerService = {
  /** POST /api/study-planners — Create a new study task. */
  create: (data: CreateStudyPlannerRequest) =>
    apiClient.post<StudyPlannerResponse>(STUDY_PLANNERS.BASE, data).then((r) => r.data),

  /** GET /api/study-planners — Get all study tasks for the authenticated user. */
  getAll: () =>
    apiClient.get<StudyPlannerResponse[]>(STUDY_PLANNERS.BASE).then((r) => r.data),

  /** GET /api/study-planners/{id} — Get a single study task by ID. */
  getById: (id: number) =>
    apiClient.get<StudyPlannerResponse>(STUDY_PLANNERS.BY_ID(id)).then((r) => r.data),

  /** PUT /api/study-planners/{id} — Update a study task. */
  update: (id: number, data: UpdateStudyPlannerRequest) =>
    apiClient.put<StudyPlannerResponse>(STUDY_PLANNERS.BY_ID(id), data).then((r) => r.data),

  /** DELETE /api/study-planners/{id} — Delete a study task. */
  delete: (id: number) =>
    apiClient.delete<string>(STUDY_PLANNERS.BY_ID(id)).then((r) => r.data),
};

File: backend/src/main/java/com/devlaunch/service/interfaces/StudyPlannerService.java
package com.devlaunch.service.interfaces;

import com.devlaunch.dto.request.CreateStudyPlannerRequest;
import com.devlaunch.dto.request.UpdateStudyPlannerRequest;
import com.devlaunch.dto.response.StudyPlannerResponse;
import com.devlaunch.exception.ResourceNotFoundException;

import java.util.List;

/**
 * Service interface for study planner management operations.
 * <p>
 * Defines the contract for creating, retrieving, updating, and deleting
 * the currently authenticated user's study planner entries. Each entry
 * represents a scheduled study session or task with a title, description,
 * date, optional time window, priority level, and completion status.
 * Update and delete operations require the entry ID and verify that
 * the entry belongs to the authenticated user.
 * </p>
 *
 * @author DevLaunch
 */
public interface StudyPlannerService {

    /**
     * Creates a new study planner entry for the currently authenticated user.
     *
     * @param request the create-study-planner request containing study session details
     * @return the newly created study planner data
     * @throws ResourceNotFoundException if the authenticated user is not found
     */
    StudyPlannerResponse createStudyPlanner(CreateStudyPlannerRequest request);

    /**
     * Retrieves all study planner entries belonging to the currently authenticated user.
     *
     * @return a list of study planner data for the authenticated user,
     *         or an empty list if none exist
     */
    List<StudyPlannerResponse> getAllStudyPlanners();

    /**
     * Retrieves a specific study planner entry by its ID, ensuring it belongs to the
     * currently authenticated user.
     *
     * @param id the study planner entry ID
     * @return the study planner data
     * @throws ResourceNotFoundException if the study planner entry is not found
     *                                    or does not belong to the user
     */
    StudyPlannerResponse getStudyPlannerById(Long id);

    /**
     * Updates a specific study planner entry by its ID, ensuring it belongs to the
     * currently authenticated user.
     *
     * @param id      the study planner entry ID to update
     * @param request the update request containing the new study session details
     * @return the updated study planner data
     * @throws ResourceNotFoundException if the study planner entry is not found
     *                                    or does not belong to the user
     */
    StudyPlannerResponse updateStudyPlanner(Long id, UpdateStudyPlannerRequest request);

    /**
     * Deletes a specific study planner entry by its ID, ensuring it belongs to the
     * currently authenticated user.
     *
     * @param id the study planner entry ID to delete
     * @throws ResourceNotFoundException if the study planner entry is not found
     *                                    or does not belong to the user
     */
    void deleteStudyPlanner(Long id);

    /**
     * Computes the current consecutive-day completion streak of the
     * authenticated user.
     * <p>
     * A streak continues from today, or from yesterday when today has no
     * completion yet, and counts backwards over consecutive days.
     * Shared with the gamification engine for the consistency badge.
     * </p>
     *
     * @return the length of the current streak in days
     */
    int getCurrentStudyStreak();

}
